mypy: Make Policies module MyPy‑compliant (#1719, #1720)

akacmazz opened the pull request:
## What this does

- Addresses #1719; fixes #1720. (🛠️ Tooling)
- Policies typing:
    - ACT: type‑safe activation helper; move batch tensors to the module device before normalization for consistent
device semantics.
    - SAC: constructor accepts None to match tests (base class raises the expected error); forward returns a typed dict.
    - Config: PreTrainedConfig now respects LEROBOT_TEST_DEVICE for deterministic device selection in CI/local runs.
- Tooling (pre‑commit): enable a scoped MyPy run via the hook (uses pyproject.toml, installs types, ignores missing
imports in the hook venv to avoid heavy deps). No global behavior change.

## How it was tested
- Static checks:
    - pre-commit run -a → all hooks green (ruff, typos, bandit, mypy).
    - mypy → success on the policies scope configured in pyproject.toml.
- Tests (CPU to keep runs deterministic and lightweight):
    - SAC: PYTHONPATH=src pytest -q tests/policies -k sac
    - ACT back‑compat (uses small artifacts): PYTHONPATH=src pytest -q tests/policies -k 'act and
backward_compatibility'
- Notes:
    - Set local HF caches to avoid permission issues and reduce flakiness.
    - On low‑memory machines, split ACT/SAC runs to avoid OOM.

## How to checkout & try? (for the reviewer)
# From repo root
export LEROBOT_TEST_DEVICE=cpu
mkdir -p .cache/hf_datasets .cache/hf_home
export HF_DATASETS_CACHE=$(pwd)/.cache/hf_datasets
export HF_HOME=$(pwd)/.cache/hf_home

# Static checks
pre-commit run -a
mypy

# SAC tests
PYTHONPATH=src pytest -q tests/policies -k sac

# ACT back-compat tests
PYTHONPATH=src pytest -q tests/policies -k 'act and backward_compatibility'
If you’d like me to extend MyPy coverage to the next module (e.g., Datasets) in a follow‑up PR, I can prepare the scoped
config and initial fixes similarly.

cc @CadeRemi @AdilZouitine for review

akacmazz committed:
mypy: make Policies MyPy-compliant (#1719, #1720)

akacmazz committed:
remove AGENTS.md

AdilZouitine commented on src/lerobot/configs/policies.py:
```diff
@@ -89,7 +94,7 @@ def __post_init__(self):
     @property
     def type(self) -> str:
-        return self.get_choice_name(self.__class__)
+        return cast(str, self.get_choice_name(self.__class__))
```
`cast` should be always justified in the code, this is unsafe. 
In this case it should be ok.

AdilZouitine commented on src/lerobot/configs/policies.py:
```diff
@@ -204,4 +210,4 @@ def from_pretrained(
             cli_overrides = policy_kwargs.pop("cli_overrides", [])
             with draccus.config_type("json"):
-                return draccus.parse(orig_config.__class__, config_file, args=cli_overrides)
+                return cast(T, draccus.parse(orig_config.__class__, config_file, args=cli_overrides))
```
Ditto: we should always justify cast. 
Question this cast is necessary?

AdilZouitine commented on src/lerobot/policies/act/modeling_act.py:
```diff
@@ -79,7 +81,7 @@ def __init__(
         self.reset()
 
-    def get_optim_params(self) -> dict:
+    def get_optim_params(self) -> object:
```
Using object here feels too generic and makes the typing less helpful. A stricter approach would be:

Option A (inline type):

`Sequence[dict[str, list[nn.Parameter] | float]]`

Option B (preferred, clearer): define a dedicated `TypedDict` for optimizer kwargs and then use:
```python
from typing import TypedDict, NotRequired
from collections.abc import Sequence
from torch import nn

class OptimizerKwargs(TypedDict, total=False):
    params: list[nn.Parameter]
    lr: float
    weight_decay: float
    momentum: float
    betas: tuple[float, float]
    eps: float

BatchKwargs = Sequence[OptimizerKwargs]
```
Option B has the advantage of being more self-documenting and makes it easier to see what fields are expected in the optimizer kwargs.

AdilZouitine commented on src/lerobot/policies/act/modeling_act.py:
```diff
@@ -136,7 +138,9 @@ def select_action(self, batch: dict[str, Tensor]) -> Tensor:
         """Predict a chunk of actions given environment observations."""
         self.eval()
-
+        # Ensure batch tensors are on the same device as the module
+        device = get_device_from_parameters(self)
+        batch = {k: (v.to(device) if torch.is_tensor(v) else v) for k, v in batch.items()}
```
This addition is not related to mypy type?

AdilZouitine commented on src/lerobot/policies/act/modeling_act.py:
```diff
@@ -146,8 +150,11 @@ def predict_action_chunk(self, batch: dict[str, Tensor]) -> Tensor:
         return actions
 
-    def forward(self, batch: dict[str, Tensor]) -> tuple[Tensor, dict]:
+    def forward(self, batch: dict[str, Tensor], **kwargs: object) -> tuple[Tensor, dict]:
```
`**kwargs` should be typed using `typing.TypedDict` and `typing_extensions.Unpack`.

AdilZouitine commented on src/lerobot/policies/act/modeling_act.py:
```diff
@@ -146,8 +150,11 @@ def predict_action_chunk(self, batch: dict[str, Tensor]) -> Tensor:
-    def forward(self, batch: dict[str, Tensor]) -> tuple[Tensor, dict]:
+    def forward(self, batch: dict[str, Tensor], **kwargs: object) -> tuple[Tensor, dict]:
         """Run the batch through the model and compute the loss for training or validation."""
+        # Ensure batch tensors are on the same device as the module
+        device = get_device_from_parameters(self)
+        batch = {k: (v.to(device) if torch.is_tensor(v) else v) for k, v in batch.items()}
```
ditto

AdilZouitine commented on src/lerobot/policies/act/modeling_act.py:
```diff
@@ -755,12 +760,12 @@ def forward(self, x: Tensor) -> Tensor:
+def get_activation_fn(activation: str) -> Callable[..., Any]:
     """Return an activation function given a string."""
     if activation == "relu":
-        return F.relu
+        return cast(Callable[..., Any], F.relu)
     if activation == "gelu":
-        return F.gelu
+        return cast(Callable[..., Any], F.gelu)
     if activation == "glu":
-        return F.glu
+        return cast(Callable[..., Any], F.glu)
```
Are `cast` necessary here?

AdilZouitine reviewed (commented):
Hey thank you for your contribution, I started the review for Act policy :smile:
I will do the next policy later 🤗

akacmazz committed:
address maintainer feedback on MyPy compliance

  - Remove --ignore-missing-imports flag from pre-commit config
  - Add TypedDict annotations (OptimizerKwargs, ForwardKwargs)
  - Improve get_optim_params() return type annotation
  - Add activation function casts for MyPy stub compatibility
  - Remove device handling code (non-MyPy related)
  - Maintain parent class method signature compatibility.

akacmazz committed:
Merge branch 'chore/mypy-policies' of https://github.com/akacmazz/lerobot into chore/mypy-policies

akacmazz committed:
Merge branch 'main' into chore/mypy-policies

AdilZouitine commented:
Hey we have merged a big pr #1452. Could you rebase this branch please ? I will finish the review 😄

zuoxingdong commented:
re-addressed by #2295
